fluent.runtime/fluent/runtime/fallback.py

```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+
+import codecs
+import os
+import six
+
+
+class FluentLocalization(object):
+    """
+    Generic API for Fluent applications.
+
+    This handles language fallback, bundle creation and string localization.
+    It uses the given resource loader to load and parse Fluent data.
+    """
+    def __init__(
+        self, locales, resource_ids, resource_loader,
+        use_isolating=False,
+        bundle_class=None, functions=None,
+    ):
+        self.locales = locales
+        self.resource_ids = resource_ids
+        self.resource_loader = resource_loader
+        self.use_isolating = use_isolating
+        if bundle_class is None:
+            from fluent.runtime import FluentBundle
+            self.bundle_class = FluentBundle
+        else:
+            self.bundle_class = bundle_class
+        self.functions = functions
+        self._bundle_cache = []
+        self._bundle_it = self._iterate_bundles()
+
+    def format_value(self, msg_id, args=None):
+        for bundle in self._bundles():
+            if not bundle.has_message(msg_id):
+                continue
+            msg = bundle.get_message(msg_id)
+            if not msg.value:
+                continue
+            val, errors = bundle.format_pattern(msg.value, args)
+            return val
+        return msg_id
+
+    def _create_bundle(self, locales):
+        return self.bundle_class(
+            locales, functions=self.functions, use_isolating=self.use_isolating
+        )
+
+    def _bundles(self):
+        bundle_pointer = 0
+        while True:
+            if bundle_pointer == len(self._bundle_cache):
+                try:
+                    self._bundle_cache.append(next(self._bundle_it))
+                except StopIteration:
+                    return
+            yield self._bundle_cache[bundle_pointer]
+            bundle_pointer += 1
+
+    def _iterate_bundles(self):
+        for first_loc in range(0, len(self.locales)):
+            locs = self.locales[first_loc:]
+            for resources in self.resource_loader.resources(locs[0], self.resource_ids):
+                bundle = self._create_bundle(locs)
+                for resource in resources:
+                    bundle.add_resource(resource)
```

Comment:
Should add the ability to pass in the `allow_override` option to `add_resource` here. Perhaps via a class property like some other options.

Reply:
`allow_override` only pays off if you do it for individual resources. Passing it to all resources is the same as just reverting the order in the loader or in the constructor.

I used this feature recently, and it was actually sub-classing the default bundle creation behavior. In terms of the code here, `_iterate_bundles` could be overwritten by a method that "prepends" a resource from a different data source, by doing something like

```python
def _iterate_bundles(self):
    for bundle in super(FunkyLocalization, self)._iterate_bundles():
        bundle.add_resource(self.bass_resource, allow_overrides=True)
        yield bundle
```